packages/mobile-app/data/wallet/writeQueue.ts

```diff
@@ -0,0 +1,178 @@
+import { Note } from "@ironfish/sdk";
+import { Network } from "../constants";
+import { WalletDb } from "./db";
+import * as UInt8ArrayUtils from "../../utils/uint8Array";
+
+enum DBWriteType {
+  ADD,
+  REMOVE,
+}
+
+type DBWrite =
+  | {
+      type: DBWriteType.ADD;
+      accountId: number;
+      sequence: number;
+      hash: Uint8Array;
+      transactions: {
+        hash: Uint8Array;
+        timestamp: Date;
+        ownerNotes: { position: number; note: Note; nullifier: string }[];
+        spenderNotes: { note: Note }[];
+        foundNullifiers: Uint8Array[];
+      }[];
+    }
+  | {
+      type: DBWriteType.REMOVE;
+      accountId: number;
+      sequence: number;
+      hash: Uint8Array;
+      prevHash: Uint8Array;
+    };
+
+/**
+ * Writing to the database for each connected block for each account has a performance impact on scanning.
+ * Typically we'll just be incrementing the account head, so we can cache the updated heads and persist them
+ * at intervals.
+ */
+export class WriteQueue {
+  readonly heads: Map<number, { hash: Uint8Array; sequence: number }> =
+    new Map();
+
+  private writeQueue: DBWrite[] = [];
+
+  readonly nullifierSet = new Set<string>();
+
+  constructor(
+    readonly db: WalletDb,
+    readonly network: Network,
+  ) {}
```

Comment:
question: Is this empty constructor intentional? If so, I'm curious what it's doing.

Reply:
If you add `readonly` to parameters in TS constructors, it automatically declares those parameters as properties on the class and assigns them in the constructor, so then the constructor body is empty. It's basically shorthand for this:

```ts
class WriteQueue {
  readonly num: number
  
  constructor(num) {
    this.num = num
  }
}
```

I don't have a style preference here, so I'll switch to the format in this comment so it's more clear.

Reply:
>  and assigns them in the constructor

Whoa, that's wild, I had no idea. My initial reaction is mixed cause it's less explicit and I was under the impression that TS doesn't mess with the logic (other than enums, which is why I'm anti-enum, personally) but I'm down to keep this since it's part of the language.

In other words, no strong feelings, so this is fine!